File: tests/00-pymod_basics/002-expect_1/test_expect_1.nim
import pymod
import unicode

template assertEqual[T](a, b: T) = doAssert(a == b)

proc cfloatExpect1*(arg: cfloat) {.exportpy.} = assertEqual(arg, 1.0)
proc cdoubleExpect1*(arg: cdouble) {.exportpy.} = assertEqual(arg, 1.0)

proc cshortExpect1*(arg: cshort) {.exportpy.} = assertEqual(arg, 1)
proc cintExpect1*(arg: cint) {.exportpy.} = assertEqual(arg, 1)
proc clongExpect1*(arg: clong) {.exportpy.} = assertEqual(arg, 1)

proc cushortExpect1*(arg: cushort) {.exportpy.} = assertEqual(arg, 1)
proc cuintExpect1*(arg: cuint) {.exportpy.} = assertEqual(arg, 1)
proc culongExpect1*(arg: culong) {.exportpy.} = assertEqual(arg, 1)

proc floatExpect1*(arg: float) {.exportpy.} = assertEqual(arg, 1.0)
proc float32Expect1*(arg: float32) {.exportpy.} = assertEqual(arg, 1.0)
proc float64Expect1*(arg: float64) {.exportpy.} = assertEqual(arg, 1.0)

proc intExpect1*(arg: int) {.exportpy.} = assertEqual(arg, 1)
# TODO
#proc int8Expect1*(arg: int8) {.exportpy.} = assertEqual(arg, 1)
proc int16Expect1*(arg: int16) {.exportpy.} = assertEqual(arg, 1)
proc int32Expect1*(arg: int32) {.exportpy.} = assertEqual(arg, 1)
proc int64Expect1*(arg: int64) {.exportpy.} = assertEqual(arg, 1)

proc uintExpect1*(arg: uint) {.exportpy.} = assertEqual(arg, 1)
proc uint8Expect1*(arg: uint8) {.exportpy.} = assertEqual(arg, 1)
proc uint16Expect1*(arg: uint16) {.exportpy.} = assertEqual(arg, 1)
proc uint32Expect1*(arg: uint32) {.exportpy.} = assertEqual(arg, 1)
proc uint64Expect1*(arg: uint64) {.exportpy.} = assertEqual(arg, 1)

# TODO
#proc boolExpect1*(arg: bool) {.exportpy.} = assertEqual(arg, True)
proc byteExpect1*(arg: byte) {.exportpy.} = assertEqual(arg, 1)

proc ccharExpect1*(arg: cchar) {.exportpy.} = assertEqual(arg, 'a')
proc charExpect1*(arg: char) {.exportpy.} = assertEqual(arg, 'a')
proc stringExpect1*(arg: string) {.exportpy.} = assertEqual(arg, "abc")

# TODO
#proc unicodeRuneExpect1*(arg: Rune) {.exportpy.} = assertEqual(arg, Rune(ord('a')))
#proc seqCharExpect1*(arg: seq[char]) {.exportpy.} = assertEqual(arg, @['a', 'b', 'c'])
#proc seqRuneExpect1*(arg: seq[Rune]) {.exportpy.} =
#  assertEqual(arg, @[Rune(ord('a')), Rune(ord('b')), Rune(ord('c'))])


initPyModule("",
    cfloatExpect1, cdoubleExpect1,
    cshortExpect1, cintExpect1, clongExpect1,
    cushortExpect1, cuintExpect1, culongExpect1,
    floatExpect1, float32Expect1, float64Expect1,
    #int8Expect1,
    intExpect1, int16Expect1, int32Expect1, int64Expect1,
    uintExpect1, uint8Expect1, uint16Expect1, uint32Expect1, uint64Expect1,
    #boolExpect1,
    byteExpect1,
    ccharExpect1, charExpect1, stringExpect1,
    #unicodeRuneExpect1, seqCharExpect1, seqRuneExpect1
    )

File: tests/00-pymod_basics/002-expect_1/test_expect_1.py
import pytest


def test_0_compile_pymod_test_mod(pmgen_py_compile):
    pmgen_py_compile(__name__)


def test_cfloatExpect1(pymod_test_mod):
    pymod_test_mod.cfloatExpect1(1.0)

def test_cdoubleExpect1(pymod_test_mod):
    pymod_test_mod.cdoubleExpect1(1.0)


def test_cshortExpect1(pymod_test_mod):
    pymod_test_mod.cshortExpect1(1)

def test_cintExpect1(pymod_test_mod):
    pymod_test_mod.cintExpect1(1)

def test_clongExpect1(pymod_test_mod):
    pymod_test_mod.clongExpect1(1)


def test_cushortExpect1(pymod_test_mod):
    pymod_test_mod.cushortExpect1(1)

def test_cuintExpect1(pymod_test_mod):
    pymod_test_mod.cuintExpect1(1)

def test_culongExpect1(pymod_test_mod):
    pymod_test_mod.culongExpect1(1)


def test_floatExpect1(pymod_test_mod):
    pymod_test_mod.floatExpect1(1.0)

def test_float32Expect1(pymod_test_mod):
    pymod_test_mod.float32Expect1(1.0)

def test_float64Expect1(pymod_test_mod):
    pymod_test_mod.float64Expect1(1.0)


def test_intExpect1(pymod_test_mod):
    pymod_test_mod.intExpect1(1)

# TODO
#def test_int8Expect1(pymod_test_mod):
#    pymod_test_mod.int8Expect1(1)

def test_int16Expect1(pymod_test_mod):
    pymod_test_mod.int16Expect1(1)

def test_int32Expect1(pymod_test_mod):
    pymod_test_mod.int32Expect1(1)

def test_int64Expect1(pymod_test_mod):
    pymod_test_mod.int64Expect1(1)


def test_uintExpect1(pymod_test_mod):
    pymod_test_mod.uintExpect1(1)

def test_uint8Expect1(pymod_test_mod):
    pymod_test_mod.uint8Expect1(1)

def test_uint16Expect1(pymod_test_mod):
    pymod_test_mod.uint16Expect1(1)

def test_uint32Expect1(pymod_test_mod):
    pymod_test_mod.uint32Expect1(1)

def test_uint64Expect1(pymod_test_mod):
    pymod_test_mod.uint64Expect1(1)


# TODO
#def test_boolExpect1(pymod_test_mod):
#    pymod_test_mod.boolExpect1(True

def test_byteExpect1(pymod_test_mod):
    pymod_test_mod.byteExpect1(1)


def test_ccharExpect1(pymod_test_mod, python_major_version):
    if python_major_version == 2:
        pymod_test_mod.ccharExpect1("a")
    else:  # Python 3 or above: bytes vs strings, yay!
        pymod_test_mod.ccharExpect1(b"a")

def test_charExpect1(pymod_test_mod, python_major_version):
    if python_major_version == 2:
        pymod_test_mod.charExpect1("a")
    else:  # Python 3 or above: bytes vs strings, yay!
        pymod_test_mod.charExpect1(b"a")

def test_stringExpect1(pymod_test_mod):
    pymod_test_mod.stringExpect1("abc")


# TODO
#def test_unicodeRuneExpect1(pymod_test_mod, python_major_version):
#    if python_major_version == 2:
#        pymod_test_mod.unicodeRuneExpect1(u"a")
#    else:  # Python 3 or above: bytes vs strings, yay!
#        pymod_test_mod.unicodeRuneExpect1("a")

# TODO
#def test_seqCharExpect1(pymod_test_mod, python_major_version):
#    if python_major_version == 2:
#        pymod_test_mod.seqCharExpect1("abc")
#    else:  # Python 3 or above: bytes vs strings, yay!
#        pymod_test_mod.seqCharExpect1(b"abc")

# TODO
#def test_seqRuneExpect1(pymod_test_mod, python_major_version):
#    if python_major_version == 2:
#        pymod_test_mod.seqRuneExpect1(u"abc")
#    else:  # Python 3 or above: bytes vs strings, yay!
#        pymod_test_mod.seqRuneExpect1("abc")


def test_floatExpect_but_supply_int(pymod_test_mod, python_major_version):
    pymod_test_mod.floatExpect1(1)

def test_floatExpect_but_supply_str(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.floatExpect1('a')
    if python_major_version == 2:
        assert str(excinfo.value) == "a float is required"
    else:  # Python 3 or above
        assert str(excinfo.value) == "a float is required"


def test_intExpect_but_supply_float(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.intExpect1(1.0)
    if python_major_version == 2:
        assert str(excinfo.value) == "integer argument expected, got float"
    else:  # Python 3 or above
        assert str(excinfo.value) == "integer argument expected, got float"

def test_intExpect_but_supply_str(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.intExpect1('a')
    if python_major_version == 2:
        assert str(excinfo.value) == "an integer is required"
    else:  # Python 3 or above
        assert str(excinfo.value) == "an integer is required (got type str)"


def test_stringExpect_but_supply_float(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.stringExpect1(1.0)
    if python_major_version == 2:
        assert str(excinfo.value) == "argument 1 must be string, not float"
    else:  # Python 3 or above
        assert str(excinfo.value) == "argument 1 must be str, not float"

def test_stringExpect_but_supply_int(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.stringExpect1(1)
    if python_major_version == 2:
        assert str(excinfo.value) == "argument 1 must be string, not int"
    else:  # Python 3 or above
        assert str(excinfo.value) == "argument 1 must be str, not int"


def test_charExpect_but_supply_str(pymod_test_mod, python_major_version):
    with pytest.raises(TypeError) as excinfo:
        pymod_test_mod.charExpect1("abc")
    if python_major_version == 2:
        assert str(excinfo.value) == "argument 1 must be char, not str"
    else:  # Python 3 or above
        assert str(excinfo.value) == "argument 1 must be a byte string of length 1, not str"

